File: packages/core/types/template/Expressions.ts
import { IBaseNode, TemplateNodeTypes } from "./BaseNode/BaseTypes";

export interface IAddressOfExpression extends IBaseNode {
  nodeType: TemplateNodeTypes.AddressOfExpression;
  type: string;
}

export interface IArraySizeAllocation extends IBaseNode {
  nodeType: TemplateNodeTypes.ArraySizeAllocation;
  length: number | string;
}

export interface IArraySubscriptExpression extends IBaseNode {
  nodeType: TemplateNodeTypes.ArraySubscriptExpression;
}

export interface IAssignmentExpression extends IBaseNode {
  nodeType: TemplateNodeTypes.AssignmentExpression;
  operator: string;
}

export interface IBinaryExpression extends IBaseNode {
  nodeType: TemplateNodeTypes.BinaryExpression;
  operator: string;
  type: string;
}

export interface ICastExpression extends IBaseNode {
  nodeType: TemplateNodeTypes.CastExpression;
  targetType: string;
}

export interface IIdentifier extends IBaseNode {
  name: string;
  nodeType: TemplateNodeTypes.Identifier;
  size?: string;
  type: string;
}

export interface ILiteral extends IBaseNode {
  nodeType: TemplateNodeTypes.Literal;
  type: string;
  size?: number;
  value: string;
}

export interface IMemberAccess extends IBaseNode {
  nodeType: TemplateNodeTypes.MemberAccess;
  type: string;
}

export interface IPointerDereference extends IBaseNode {
  nodeType: TemplateNodeTypes.PointerDereference;
}

export interface ISizeOfExpression extends IBaseNode {
  nodeType: TemplateNodeTypes.SizeOfExpression;
}

export interface IStandardLibCall extends IBaseNode {
  name: string;
  nodeType: TemplateNodeTypes.StandardLibCall;
}

export interface IUnaryExpression extends IBaseNode {
  nodeType: TemplateNodeTypes.UnaryExpression;
  operator: string;
  type: string;
}

export interface IUserDefinedCall extends IBaseNode {
  nodeType: TemplateNodeTypes.UserDefinedCall;
}

File: py-packages/core/src/core/types/template/expressions.py
"""Expression node type definitions."""

from typing import Optional

from pydantic import BaseModel

from .BaseNode.base_types import IBaseNode, TemplateNodeTypes


class IAddressOfExpression(IBaseNode):
    """Address-of expression node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.AddressOfExpression
    type: str


class IArraySizeAllocation(IBaseNode):
    """Array size allocation node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.ArraySizeAllocation
    length: int | str


class IArraySubscriptExpression(IBaseNode):
    """Array subscript expression node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.ArraySubscriptExpression


class IAssignmentExpression(IBaseNode):
    """Assignment expression node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.AssignmentExpression
    operator: str


class IBinaryExpression(IBaseNode):
    """Binary expression node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.BinaryExpression
    operator: str
    type: str


class ICastExpression(IBaseNode):
    """Cast expression node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.CastExpression
    targetType: str


class IIdentifier(IBaseNode):
    """Identifier node."""

    name: str
    nodeType: TemplateNodeTypes = TemplateNodeTypes.Identifier
    size: Optional[str] = None
    type: str


class ILiteral(IBaseNode):
    """Literal node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.Literal
    type: str
    size: Optional[int] = None
    value: str


class IMemberAccess(IBaseNode):
    """Member access node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.MemberAccess
    type: str


class IPointerDereference(IBaseNode):
    """Pointer dereference node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.PointerDereference


class ISizeOfExpression(IBaseNode):
    """Sizeof expression node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.SizeOfExpression


class IStandardLibCall(IBaseNode):
    """Standard library call node."""

    name: str
    nodeType: TemplateNodeTypes = TemplateNodeTypes.StandardLibCall


class IUnaryExpression(IBaseNode):
    """Unary expression node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.UnaryExpression
    operator: str
    type: str


class IUserDefinedCall(IBaseNode):
    """User-defined call node."""

    nodeType: TemplateNodeTypes = TemplateNodeTypes.UserDefinedCall


